### ISSUE TITLE
New command: dnu sources

### ISSUE THREAD

**Issue by victorhurdugaci** [aspnet/Announcements#46]
**TL;DR;** 

`dnu sources <package-name>` will retrieve the source code for a the package named `<package-name>`

The feature is in a very early stage and we've rolled it out to get feedback. Please let us know what you think. See Limitations at the end of this post.

-----

A few months ago I wrote [an article about debugging ASP.NET 5 framework code](http://blogs.msdn.com/b/webdev/archive/2015/02/06/debugging-asp-net-5-framework-code-using-visual-studio-2015.aspx). While not impossible, it is hard to determine which commit produced a particular package if the commit was not tagged.

Today, we are excited to announce `dnu sources`, a new feature that will simplify the process of getting source files for packages.

`dnu sources` allows you to get the source code for a particular package, without you having to look online for it.

### How to add repository information to a package

In order for a package to support `dnu sources` it must include some new information in the `project.json` file. 

```
"repository": { 
    "type": "git", 
    "url": "https://github.com/aspnet/options" 
}
```

[source](https://github.com/aspnet/Options/blob/dev/src/Microsoft.Framework.OptionsModel/project.json#L9-L12)

In addition to the new `repository` property, the following conditions must be satisfied:

1. The source code must be in a Git repository at build time
2. Git must be installed on the machine and added to the PATH
 
The embed repository information, simply build the package (`dnu pack`),

### How to get the sources for a package 

If a package has source information, you will be able to run `dnu sources <package>` name to retrieve the sources for `<package>`.

Conditions for running `dnu sources`

1. The project must have a reference to package you are trying to source.
2. The package must be installed locally

### Try it out

1. Add the following reference to `project.json`: `Microsoft.Framework.OptionsModel`.
2. Restore the packages using Visual Studio or by running `dnu restore`
3. In the project's folder, run `dnu sources Microsoft.Framework.OptionsModel`

### Limitations

1. If you don't have the dev feed (myget.org/aspnetvnext) as a NuGet source and the package references any of the `*.Sources` packages (like `Microsoft.Framework.NotNullAttribute.Sources`), you will not be able to compile the new sources. The workaround is to add the feed to NuGet.config. 
2. If you application references multiple versions of the same package, you will not be able to use `dnu source`
3. Each version of a package will create a new folder under `%userprofile%\.dnx\sources`. That folder is never cleaned up.
4. Currently, only the packages shipped by the ASP.NET and EF teams have source information. The CLR packages (System*), Roslyn or any other packages does not have source information. You will get an error message if you try to source them.
5. Only Git is supported.
6. There is no tooling support.
7. `dnu sources` works only for packages produced in `beta6` or later. Any `beta5` or earlier packages do not have source information in them.

**Comment by eashi** [comment]
Wonnnderful news!

**Comment by avanderhoorn** [comment]
Is the repository/commit info accessible at runtime (i.e. via ILibManager maybe) and if so can you provide an example of accessing it?

**Comment by victorhurdugaci** [comment]
@avanderhoorn no, it is not available but it can be added. What's the scenario in which you need it?

**Comment by avanderhoorn** [comment]
Trying to pull that information out within Glimpse to show them which commit they are on, etc. Also we have plans to show which packages are in play, and for the packages which had this information available, we could link through to the repo at that commit. A nice to have would be if that commit had any tag information associated with it as well.

**Comment by victorhurdugaci** [comment]
@avanderhoorn https://github.com/aspnet/dnx/issues/2349

We can expose the commit id and the repo url (the information that `dnu sources` uses). For the tag, your code will have to do a query.

**Comment by mrtaikandi** [comment]
Awesome.

**Comment by bradwilson** [comment]
Is there any intention of being able to provide the necessary values rather than presume a Git repository? Some build systems pull source and push it over to the build machine, rather than cloning it directly on the box.

**Comment by davidfowl** [comment]
@bradwilson We haven't done thought about that as yet but it would certainly be doable with an env variable. Are you doing that with your CI?

**Comment by bradwilson** [comment]
It's not just agent-side checkout, but also installing Git onto the build machines.

**Comment by JamesNK** [comment]
Will this only be supported via project.json and not the old xml NuGet spec file?

**Comment by davidfowl** [comment]
@JamesNK we spoke with the NuGet team about this going into the nuspec. It's not supported there as yet because we wanted to do a trial run of this feature first. It's also very tightly coupled to dnx style projects. Existing projects have a more loose source layout and would likely need alot more information for "clone as source" to just work with minimal changes.

**Comment by NeelBhatt** [comment]
It is great!

**Comment by NeelBhatt** [comment]
Great! I have wrote blog on this here : https://neelbhatt40.wordpress.com/2015/08/07/get-source-code-via-dnu-sources-a-new-born-mvc-command/

**Comment by victorhurdugaci** [comment]
@NeelBhatt awesome!

Everyone, I'm locking the thread. Please use [this dnx thread](https://github.com/aspnet/dnx/issues/2437) for further discussions about dnu sources.
